include/deal.II/base/table_indices.h

```diff
@@ -43,11 +45,71 @@ class TableIndices
 public:
 
   /**
-   * Default constructor. It sets all indices to zero.
+   * Default constructor. This constructor sets all indices to zero.
    */
   TableIndices();
 
   /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<1> and initializes the single
+   * index with @p index0.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from one.
+   */
+  explicit TableIndices (const unsigned int index0);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<2> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from two.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<3> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from three.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<4> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from four.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<5> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from five.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3,
+                const unsigned int index4);
```

Comment:
Hmm... So what types do you want to adress saying  "given type could be converted to an unsigned int in a function argument list without warning"? I guess std::is_integral looks reasonable, doesn't it? Then, I would be in favor of the follwoing
```
#include <iostream>
#include <array>

template <typename...> struct all_integral;

template <> struct all_integral<>
{
  static constexpr bool value = true;
};

template <typename T, typename... Rest> struct all_integral<T, Rest...>
{
  static constexpr bool value = std::is_integral<T>::value
                                && all_integral<Rest...>::value;
};

template <int N>
struct TableIndices
{
  template <typename... T>
  explicit TableIndices (const T... args)
    : values{{static_cast<const unsigned int>(args)...}}
    {
      static_assert (all_integral<T...>::value,
                     "Not all of the parameters have integral type!");
      static_assert (sizeof...(T) == N,
                     "Wrong number of constructor arguments!");
    }

private:
  std::array<unsigned int, N> values;
};
```